### ISSUE TITLE
pydantic warning Field "model_type" has conflict with protected namespace "model_".

### ISSUE THREAD

**Issue by ehartford** [axolotl-ai-cloud/axolotl#1335]
### Please check that this issue hasn't been reported before.

- [X] I searched previous [Bug Reports](https://github.com/OpenAccess-AI-Collective/axolotl/labels/bug) didn't find any similar reports.

### Expected Behavior

no pydantic warnings

### Current behaviour

/home/ubuntu/miniconda3/envs/axolotl/lib/python3.11/site-packages/pydantic/_internal/_fields.py:151: UserWarning: Field "model_type" has conflict with protected namespace "model_".

### Steps to reproduce

train model

### Config yaml

```yaml
base_model: google/gemma-7b
model_type: AutoModelForCausalLM
tokenizer_type: AutoTokenizer

load_in_8bit: false
load_in_4bit: true
strict: false

datasets:
  - path: /home/ubuntu/datasets/dolphin-coder-unsloth/dolphin201-sharegpt2.jsonl
    type: sharegpt
  - path: /home/ubuntu/datasets/dolphin-coder-unsloth/dolphin-coder-translate-sharegpt2.jsonl
    type: sharegpt
  - path: /home/ubuntu/datasets/dolphin-coder-unsloth/dolphin-coder-codegen-sharegpt2.jsonl
    type: sharegpt
  - path: /home/ubuntu/datasets/dolphin-coder-unsloth/data-evol_instruct-sharegpt2.jsonl
    type: sharegpt
  - path: /home/ubuntu/datasets/dolphin-coder-unsloth/data-oss_instruct-decontaminated-sharegpt2.jsonl
    type: sharegpt
  - path: /home/ubuntu/datasets/not_samantha_norefusals.jsonl
    type: sharegpt
  - path: /home/ubuntu/datasets/openhermes2_5-sharegpt.jsonl
    type: sharegpt

val_set_size: 0
output_dir: /home/ubuntu/dolphin-2.8-gemma-7b

adapter: qlora
lora_r: 32
lora_alpha: 16
lora_dropout: 0.05
lora_target_linear: true

sequence_len: 8192
sample_packing: true
pad_to_sequence_len: true

wandb_project: dolphin
wandb_entity:
wandb_watch:
wandb_name:
wandb_log_model:

gradient_accumulation_steps: 8
micro_batch_size: 2
num_epochs: 3
optimizer: adamw_bnb_8bit
lr_scheduler: cosine
learning_rate: 2e-4

train_on_inputs: false
group_by_length: false
bf16: auto
fp16:
tf32: false

gradient_checkpointing: true
early_stopping_patience:
resume_from_checkpoint:
local_rank:
logging_steps: 1
xformers_attention:
flash_attention: true

warmup_steps: 10
evals_per_epoch: 4
eval_table_size:
eval_max_new_tokens: 128
saves_per_epoch: 2
debug:
deepspeed: deepspeed_configs/zero2.json
weight_decay: 0.0
fsdp:
fsdp_config:
```

### Which Operating Systems are you using?

- [X] Linux
- [ ] macOS
- [ ] Windows

### Python Version

3.11

### axolotl branch-commit

main

### Acknowledgements

- [X] My issue title is concise, descriptive, and in title casing.
- [X] I have searched the existing issues to make sure this bug has not been reported yet.
- [X] I am using the latest version of axolotl.
- [X] I have provided enough information for the maintainers to reproduce and diagnose the issue.

**Comment by semantium** [comment]
I'm also having this issue

**Comment by Jay-Nehra** [comment]
**I am facing the same issue as well. Issue is causing the training to break at the `Gradient Checkpoint` after the Epoch is completed . Training loop break when the first epoch is completed. 

Issue seems to be easily reproducible as well. I had the same error after upgrading both `Axolotl` and `Pydantic`.**  

Here are the warnings before the training begins :

> `/usr/local/lib/python3.10/dist-packages/pydantic/_internal/_fields.py:151: UserWarning: Field "model_type" has conflict with protected namespace "model_".
> 
> You may be able to resolve this warning by setting `model_config['protected_namespaces'] = ()`.
>   warnings.warn(
> /usr/local/lib/python3.10/dist-packages/pydantic/_internal/_fields.py:151: UserWarning: Field "model_revision" has conflict with protected namespace "model_".
> 
> You may be able to resolve this warning by setting `model_config['protected_namespaces'] = ()`.
>   warnings.warn(
> /usr/local/lib/python3.10/dist-packages/pydantic/_internal/_fields.py:151: UserWarning: Field "model_config_overrides" has conflict with protected namespace "model_".
> 
> You may be able to resolve this warning by setting `model_config['protected_namespaces'] = ()`.
>   warnings.warn(
> /usr/local/lib/python3.10/dist-packages/pydantic/_internal/_fields.py:151: UserWarning: Field "model_type" has conflict with protected namespace "model_".
> 
> You may be able to resolve this warning by setting `model_config['protected_namespaces'] = ()`.
>   warnings.warn(
> /usr/local/lib/python3.10/dist-packages/pydantic/_internal/_fields.py:151: UserWarning: Field "model_revision" has conflict with protected namespace "model_".
> 
> You may be able to resolve this warning by setting `model_config['protected_namespaces'] = ()`.
>   warnings.warn(
> /usr/local/lib/python3.10/dist-packages/pydantic/_internal/_fields.py:151: UserWarning: Field "model_config_overrides" has conflict with protected namespace "model_".
> 
> You may be able to resolve this warning by setting `model_config['protected_namespaces'] = ()`.
>   warnings.warn(
> `

**Comment by Jay-Nehra** [comment]
#### Config YAML

> base_model: mistralai/Mistral-7B-Instruct-v0.2
> model_type: MistralForCausalLM
> tokenizer_type: LlamaTokenizer
> is_mistral_derived_model: true
> 
> load_in_8bit: true
> load_in_4bit: false
> strict: false
> 
> datasets:
>   - path: jsonl_format_dataset
>     ds_type: json
>     type:
>       system_format: "<<SYS>> {system} <</SYS>> "
>       format: "[INST] {input} [/INST]"
>       field_system: system
>       field_instruction: input
>       field_output: output
> 
> dataset_prepared_path: 
> val_set_size: 0.05
> output_dir: ./experimental_finetuned_model_v1.2
> 
> adapter: lora
> lora_r: 8
> lora_alpha: 16
> lora_dropout: 0.05
> lora_target_modules:
>   - q_proj
>   - v_proj
> 
> sequence_len: 512
> sample_packing: false
> pad_to_sequence_len: true
> 
> wandb_project:
> wandb_entity:
> wandb_watch:
> wandb_name:
> wandb_log_model:
> 
> gradient_accumulation_steps: 3
> micro_batch_size: 2
> num_epochs: 10
> optimizer: adamw_bnb_8bit
> lr_scheduler: cosine
> learning_rate: 0.0002
> 
> train_on_inputs: false
> group_by_length: false
> bf16: true
> fp16: false
> tf32: false
> 
> gradient_checkpointing: true
> early_stopping_patience: 
> resume_from_checkpoint: 
> local_rank:
> logging_steps: 1
> xformers_attention:
> flash_attention: true
> 
> warmup_steps: 10
> evals_per_epoch: 4
> eval_table_size:
> eval_max_new_tokens: 128
> saves_per_epoch: 1
> debug:
> 
> deepspeed: /workspace/axolotl/deepspeed_configs/zero1.json
> weight_decay: 0.0
> fsdp:
> fsdp_config:
> special_tokens:
>   bos_token: "<s>"
>   eos_token: "</s>"
>   unk_token: "<unk>"

**Comment by winglian** [comment]
This should be fixed in the latest main by #1345

**Comment by Jay-Nehra** [comment]
Hi @winglian, 

I tried running the trainer again after taking the fresh pull. `Pydantic Model Warning` is **resolved**. **Thanks for this**. 

But The Second Issue that was happening at the `Gradient Checkpoint ` is I think due to  PyTorch Update.  I am not able to debug the error further right now and I am not sure if there is some other issue that was already opened on this so I logged my issue here as well (I'll see if there is something in the `open issues`  else I'll open the issue in couple of hours.). 

Our training loop is still breaking after the first Epoch. 

I will check the open issues in some time to look for a more relevant Issue. Positional Agument use has been deprecated and that is causing the issue. 

ps: Here is the error log :
```

>  10%|█████████▎                                                                                   | 182/1820 [11:32<1:35:56,  3.51s/it]/usr/local/lib/python3.10/dist-packages/torch/nn/modules/module.py:1879: UserWarning: Positional args are being deprecated, use kwargs instead. Refer to https://pytorch.org/docs/master/generated/torch.nn.Module.html#torch.nn.Module.state_dict for details.
>   warnings.warn(
> /usr/local/lib/python3.10/dist-packages/torch/nn/modules/module.py:1879: UserWarning: Positional args are being deprecated, use kwargs instead. Refer to https://pytorch.org/docs/master/generated/torch.nn.Module.html#torch.nn.Module.state_dict for details.
>   warnings.warn(
> Traceback (most recent call last):
>   File "/usr/lib/python3.10/runpy.py", line 196, in _run_module_as_main
>     return _run_code(code, main_globals, None,
>   File "/usr/lib/python3.10/runpy.py", line 86, in _run_code
>     exec(code, run_globals)
>   File "/workspace/axolotl/src/axolotl/cli/train.py", line 59, in <module>
>     fire.Fire(do_cli)
>   File "/usr/local/lib/python3.10/dist-packages/fire/core.py", line 141, in Fire
>     component_trace = _Fire(component, args, parsed_flag_args, context, name)
>   File "/usr/local/lib/python3.10/dist-packages/fire/core.py", line 475, in _Fire
>     component, remaining_args = _CallAndUpdateTrace(
>   File "/usr/local/lib/python3.10/dist-packages/fire/core.py", line 691, in _CallAndUpdateTrace
>     component = fn(*varargs, **kwargs)
>   File "/workspace/axolotl/src/axolotl/cli/train.py", line 35, in do_cli
>     return do_train(parsed_cfg, parsed_cli_args)
>   File "/workspace/axolotl/src/axolotl/cli/train.py", line 55, in do_train
>     return train(cfg=cfg, cli_args=cli_args, dataset_meta=dataset_meta)
>   File "/workspace/axolotl/src/axolotl/train.py", line 163, in train
>     trainer.train(resume_from_checkpoint=resume_from_checkpoint)
>   File "/usr/local/lib/python3.10/dist-packages/transformers/trainer.py", line 1624, in train
>     return inner_training_loop(
>   File "/usr/local/lib/python3.10/dist-packages/transformers/trainer.py", line 2029, in _inner_training_loop
>     self._maybe_log_save_evaluate(tr_loss, grad_norm, model, trial, epoch, ignore_keys_for_eval)
>   File "/usr/local/lib/python3.10/dist-packages/transformers/trainer.py", line 2423, in _maybe_log_save_evaluate
>     self._save_checkpoint(model, trial, metrics=metrics)
>   File "/usr/local/lib/python3.10/dist-packages/transformers/trainer.py", line 2525, in _save_checkpoint
>     self.state.save_to_json(os.path.join(staging_output_dir, TRAINER_STATE_NAME))
>   File "/usr/local/lib/python3.10/dist-packages/transformers/trainer_callback.py", line 113, in save_to_json
>     json_string = json.dumps(dataclasses.asdict(self), indent=2, sort_keys=True) + "\n"
>   File "/usr/lib/python3.10/json/__init__.py", line 238, in dumps
>     **kw).encode(obj)
>   File "/usr/lib/python3.10/json/encoder.py", line 201, in encode
>     chunks = list(chunks)
>   File "/usr/lib/python3.10/json/encoder.py", line 431, in _iterencode
>     yield from _iterencode_dict(o, _current_indent_level)
>   File "/usr/lib/python3.10/json/encoder.py", line 405, in _iterencode_dict
>     yield from chunks
>   File "/usr/lib/python3.10/json/encoder.py", line 325, in _iterencode_list
>     yield from chunks
>   File "/usr/lib/python3.10/json/encoder.py", line 405, in _iterencode_dict
>     yield from chunks
>   File "/usr/lib/python3.10/json/encoder.py", line 438, in _iterencode
>     o = _default(o)
>   File "/usr/lib/python3.10/json/encoder.py", line 179, in default
>     raise TypeError(f'Object of type {o.__class__.__name__} '
> TypeError: Object of type Tensor is not JSON serializable
>  10%|█████████▎                                                                                   | 182/1820 [11:36<1:44:26,  3.83s/it]
> [2024-02-29 03:45:30,180] torch.distributed.elastic.multiprocessing.api: [WARNING] Sending process 371643 closing signal SIGTERM
> [2024-02-29 03:45:30,796] torch.distributed.elastic.multiprocessing.api: [ERROR] failed (exitcode: 1) local_rank: 0 (pid: 371642) of binary: /usr/bin/python
> Traceback (most recent call last):
>   File "/usr/local/bin/accelerate", line 8, in <module>
>     sys.exit(main())
>   File "/usr/local/lib/python3.10/dist-packages/accelerate/commands/accelerate_cli.py", line 47, in main
>     args.func(args)
>   File "/usr/local/lib/python3.10/dist-packages/accelerate/commands/launch.py", line 1014, in launch_command
>     multi_gpu_launcher(args)
>   File "/usr/local/lib/python3.10/dist-packages/accelerate/commands/launch.py", line 672, in multi_gpu_launcher
>     distrib_run.run(args)
>   File "/usr/local/lib/python3.10/dist-packages/torch/distributed/run.py", line 797, in run
>     elastic_launch(
>   File "/usr/local/lib/python3.10/dist-packages/torch/distributed/launcher/api.py", line 134, in __call__
>     return launch_agent(self._config, self._entrypoint, list(args))
>   File "/usr/local/lib/python3.10/dist-packages/torch/distributed/launcher/api.py", line 264, in launch_agent
>     raise ChildFailedError(
> torch.distributed.elastic.multiprocessing.errors.ChildFailedError: 
> ============================================================
> axolotl.cli.train FAILED
> ------------------------------------------------------------
> Failures:
>   <NO_OTHER_FAILURES>
> ------------------------------------------------------------
> Root Cause (first observed failure):
> [0]:
>   time      : 2024-02-29_03:45:30
>   host      : 2fd313022f21
>   rank      : 0 (local_rank: 0)
>   exitcode  : 1 (pid: 371642)
>   error_file: <N/A>
>   traceback : To enable traceback see: https://pytorch.org/docs/stable/elastic/errors.html

```

**Comment by Jay-Nehra** [comment]
Hello @winglian , 

This is happening due to dependency conflicts.

**Comment by winglian** [comment]
@Jay-Nehra What dependency conflicts? Are you using deepspeed? if so what version? see https://github.com/OpenAccess-AI-Collective/axolotl/issues/1320#issuecomment-1962329372

**Comment by winglian** [comment]
@Jay-Nehra there is already a PR upstream to fix this https://github.com/huggingface/transformers/pull/29212

**Comment by Jay-Nehra** [comment]
> @Jay-Nehra there is already a PR upstream to fix this [huggingface/transformers#29212](https://github.com/huggingface/transformers/pull/29212)

@winglian  This PR addresses the issue.
